Use a bytes.Buffer for shell_words string concat

dnephin opened the pull request:
Concatenating strings with a `bytes.Buffer` is much faster than `+=`

Added some `switch/case` and an early return guard for readability.
Updated a couple error messages to use `pkg/errors` and lowercase letters.

dnephin committed:
Use a bytes.Buffer for shell_words string concat

It's much faster

Signed-off-by: Daniel Nephin <dnephin@docker.com>

aaronlehmann commented on builder/dockerfile/shell_parser.go:
```diff
@@ -7,7 +7,8 @@ package dockerfile
 import (
-	"fmt"
+	"bytes"
+	"github.com/pkg/errors"
```
Move below with a blank line between this and the standard library imports

aaronlehmann commented on builder/dockerfile/shell_parser.go:
```diff
@@ -203,127 +201,122 @@ func (sw *shellWord) processDoubleQuote() (string, error) {
 				chNext := sw.scanner.Peek()
-
-				if chNext == scanner.EOF {
+				switch chNext {
```
`chNext` is never used elsewhere. This could be `switch sw.scanner.Peek() {` (but it doesn't matter).

aaronlehmann commented:
LGTM

duglin commented on builder/dockerfile/shell_parser.go:
```diff
@@ -169,25 +170,23 @@ func (sw *shellWord) processSingleQuote() (string, error) {
 
 	sw.scanner.Next()
 
+loop:
```
I know its a personal preference thing, but I'm really not a fan of this type of `break` - it just feels too much like a goto to me.

duglin replied:
And if the only reason we're doing it is to switch between an IF and a CASE I don't see the benefit.

dnephin replied:
Yes, it rubs me a bit the wrong way as well, but I think it might be an unfortunate necessity in golang.

While it looks like a GOTO, I don't think it acts like one. It still acts like a regular break statement, it just gives you a way to target the "thing" you're breaking out of.

An alternative is to just return here instead. Nothing is done after this break, so no other code changes are required. I went back and forth between these options and thought I'd try this first.

I'm fine to change this to an early return.

vdemeester replied:
I'm with @duglin on that one 😛 I find that one really confusing 👼

tonistiigi reviewed (approved):
LGTM

dnephin commented:
Cool, I removed the break/label and replaced it with a return. It turns out that the line isn't duplicated after all. It just returns from the single case now.

Good call!

vdemeester reviewed (approved):
LGTM 🐯

duglin commented:
thanks LGTM once janky is happy
